# Churn case Part2 Analysis of LTV

In [171]:
import sqlite3
import pandas as pd

In [172]:
conn =sqlite3.connect("churn.db")

In [173]:
df= pd.read_sql("select * from churn_all",conn)

In [174]:
df.head()

CustomerID  Gender SeniorCitizen Partner Dependents       State   Latitude  \
0  3668-QPYBK    Male            No      No         No  California  33.964131   
1  9237-HQITU  Female            No      No        Yes  California  34.059281   
2  9305-CDSKC  Female            No      No        Yes  California  34.048013   
3  7892-POOKP  Female            No     Yes        Yes  California  34.062125   
4  0280-XJGEX    Male            No      No        Yes  California  34.039224   

    Longitude  ZipCode PhoneService  ... TechSupport StreamingTV  \
0 -118.272783    90003          Yes  ...          No          No   
1 -118.307420    90005          Yes  ...          No          No   
2 -118.293953    90006          Yes  ...          No         Yes   
3 -118.315709    90010          Yes  ...         Yes         Yes   
4 -118.266293    90015          Yes  ...          No         Yes   

  StreamingMovies Tenure        Contract              PaymentMethod  \
0              No      2  Month-to-month               Mailed check   
1              No      2  Month-to-month           Electronic check   
2             Yes      8  Month-to-month           Electronic check   
3             Yes     28  Month-to-month           Electronic check   
4             Yes     49  Month-to-month  Bank transfer (automatic)   

  PaperlessBilling MonthlyCharges  TotalCharges Churn  
0              Yes          53.85        108.15   Yes  
1              Yes          70.70        151.65   Yes  
2              Yes          99.65         820.5   Yes  
3              Yes         104.80       3046.05   Yes  
4              Yes         103.70        5036.3   Yes  

[5 rows x 25 columns]

## 1.For those customers who unsubscribed the service, what was their average LTV(Lifetime Value)? and how long did they usually stay in the service?

In [175]:
## Extract those who unsubscribed the service( Churn = Yes)

churn_df= pd.read_sql("select * from churn_all where Churn ='Yes'",conn)

In [207]:
# Examine the distribution of Total charges

churn_df["TotalCharges"].describe()

count    1869.000000
mean     1531.796094
std      1890.822994
min        18.850000
25%       134.500000
50%       703.550000
75%      2331.300000
max      8684.800000
Name: TotalCharges, dtype: float64

Around 20 % of customer's Total Charges are very high

Divide the data to see distribution

In [177]:
# Find the 80th percentile of the data in TotalCharges
churn_df.TotalCharges.quantile(0.8)

2840.4100000000003

In [178]:
pd.read_sql("select * from churn_all where TotalCharges <=2840.41",conn)

CustomerID  Gender SeniorCitizen Partner Dependents       State  \
0     3668-QPYBK    Male            No      No         No  California   
1     9237-HQITU  Female            No      No        Yes  California   
2     9305-CDSKC  Female            No      No        Yes  California   
3     4190-MFLUW  Female            No     Yes         No  California   
4     8779-QRDMV    Male           Yes      No         No  California   
...          ...     ...           ...     ...        ...         ...   
4734  8456-QDAVC    Male            No      No         No  California   
4735  7750-EYXWZ  Female            No      No         No  California   
4736  2569-WGERO  Female            No      No         No  California   
4737  6840-RESVB    Male            No     Yes        Yes  California   
4738  4801-JZAZL  Female            No     Yes        Yes  California   

       Latitude   Longitude  ZipCode PhoneService  ...          TechSupport  \
0     33.964131 -118.272783    90003          Yes  ...                   No   
1     34.059281 -118.307420    90005          Yes  ...                   No   
2     34.048013 -118.293953    90006          Yes  ...                   No   
3     34.066367 -118.309868    90020          Yes  ...                  Yes   
4     34.023810 -118.156582    90022           No  ...                   No   
...         ...         ...      ...          ...  ...                  ...   
4734  32.852947 -114.850784    92283          Yes  ...                   No   
4735  34.159534 -116.425984    92284           No  ...                  Yes   
4736  34.341737 -116.539416    92285          Yes  ...  No internet service   
4737  34.667815 -117.536183    92301          Yes  ...                  Yes   
4738  34.167800 -116.864330    92305           No  ...                   No   

              StreamingTV      StreamingMovies Tenure        Contract  \
0                      No                   No      2  Month-to-month   
1                      No                   No      2  Month-to-month   
2                     Yes                  Yes      8  Month-to-month   
3                      No                   No     10  Month-to-month   
4                      No                  Yes      1  Month-to-month   
...                   ...                  ...    ...             ...   
4734                  Yes                   No     19  Month-to-month   
4735                  Yes                  Yes     12        One year   
4736  No internet service  No internet service     72        Two year   
4737                  Yes                  Yes     24        One year   
4738                   No                   No     11  Month-to-month   

                  PaymentMethod PaperlessBilling MonthlyCharges  TotalCharges  \
0                  Mailed check              Yes          53.85        108.15   
1              Electronic check              Yes          70.70        151.65   
2              Electronic check              Yes          99.65        820.50   
3       Credit card (automatic)               No          55.20        528.35   
4              Electronic check              Yes          39.65         39.65   
...                         ...              ...            ...           ...   
4734  Bank transfer (automatic)              Yes          78.70       1495.10   
4735           Electronic check               No          60.65        743.30   
4736  Bank transfer (automatic)              Yes          21.15       1419.40   
4737               Mailed check              Yes          84.80       1990.50   
4738           Electronic check              Yes          29.60        346.45   

     Churn  
0      Yes  
1      Yes  
2      Yes  
3      Yes  
4      Yes  
...    ...  
4734    No  
4735    No  
4736    No  
4737    No  
4738    No  

[4739 rows x 25 columns]

In [179]:
churn_df.query("TotalCharges <=2840.41")

CustomerID  Gender SeniorCitizen Partner Dependents       State  \
0     3668-QPYBK    Male            No      No         No  California   
1     9237-HQITU  Female            No      No        Yes  California   
2     9305-CDSKC  Female            No      No        Yes  California   
5     4190-MFLUW  Female            No     Yes         No  California   
6     8779-QRDMV    Male           Yes      No         No  California   
...          ...     ...           ...     ...        ...         ...   
1863  1122-JWTJW    Male            No     Yes        Yes  California   
1864  1699-HPSBG    Male            No      No         No  California   
1865  8775-CEBBJ  Female            No      No         No  California   
1866  6894-LFHLY    Male           Yes      No         No  California   
1868  8361-LTMKD    Male           Yes     Yes         No  California   

       Latitude   Longitude  ZipCode PhoneService  ... TechSupport  \
0     33.964131 -118.272783    90003          Yes  ...          No   
1     34.059281 -118.307420    90005          Yes  ...          No   
2     34.048013 -118.293953    90006          Yes  ...          No   
5     34.066367 -118.309868    90020          Yes  ...         Yes   
6     34.023810 -118.156582    90022           No  ...          No   
...         ...         ...      ...          ...  ...         ...   
1863  32.698964 -115.886656    92259          Yes  ...          No   
1864  33.745746 -116.514215    92264          Yes  ...         Yes   
1865  32.790282 -115.689559    92273          Yes  ...          No   
1866  34.264124 -114.717964    92280          Yes  ...          No   
1868  34.550820 -117.122396    92307          Yes  ...          No   

     StreamingTV StreamingMovies Tenure        Contract  \
0             No              No      2  Month-to-month   
1             No              No      2  Month-to-month   
2            Yes             Yes      8  Month-to-month   
5             No              No     10  Month-to-month   
6             No             Yes      1  Month-to-month   
...          ...             ...    ...             ...   
1863          No              No      1  Month-to-month   
1864         Yes              No     12        One year   
1865          No              No      9  Month-to-month   
1866          No              No      1  Month-to-month   
1868          No              No      4  Month-to-month   

                  PaymentMethod PaperlessBilling MonthlyCharges  TotalCharges  \
0                  Mailed check              Yes          53.85        108.15   
1              Electronic check              Yes          70.70        151.65   
2              Electronic check              Yes          99.65        820.50   
5       Credit card (automatic)               No          55.20        528.35   
6              Electronic check              Yes          39.65         39.65   
...                         ...              ...            ...           ...   
1863               Mailed check              Yes          70.65         70.65   
1864           Electronic check              Yes          59.80        727.80   
1865  Bank transfer (automatic)              Yes          44.20        403.35   
1866           Electronic check              Yes          75.75         75.75   
1868               Mailed check              Yes          74.40        306.60   

     Churn  
0      Yes  
1      Yes  
2      Yes  
5      Yes  
6      Yes  
...    ...  
1863   Yes  
1864   Yes  
1865   Yes  
1866   Yes  
1868   Yes  

[1495 rows x 25 columns]

In [180]:
# Divide the data by the 80th percentile of the data from the TotalCharges variable
total_charges_under80 =churn_df.query("TotalCharges <=2840.41")
total_charges_above80 =churn_df.query("TotalCharges >2840.41")

In [181]:
# Show the distribution of people under the 80th percentile
total_charges_under80.TotalCharges.describe()

count    1495.000000
mean      711.265819
std       766.848197
min        18.850000
25%        85.025000
50%       371.650000
75%      1128.225000
max      2839.650000
Name: TotalCharges, dtype: float64

In [206]:
# Show the distribution of people under the 80th percentile
total_charges_above80.TotalCharges.describe()

count     374.000000
mean     4811.723262
std      1436.724288
min      2841.550000
25%      3522.462500
50%      4571.100000
75%      5891.212500
max      8684.800000
Name: TotalCharges, dtype: float64

In [183]:
# Show the distribution of the tenure of people under the 80th percentile
total_charges_under80.Tenure.describe()

count    1495.000000
mean        9.935117
std        10.742349
min         1.000000
25%         1.000000
50%         6.000000
75%        15.000000
max        61.000000
Name: Tenure, dtype: float64

In [184]:
# Show the distribution of the tenure of people above the 80th percentile
total_charges_above80.Tenure.describe()

count    374.000000
mean      50.133690
std       12.334841
min       27.000000
25%       40.000000
50%       49.500000
75%       60.000000
max       72.000000
Name: Tenure, dtype: float64

**Ans:** 

-The average LTV of the **80% of those who unsubscribed is $711, and their     tenure is near 10 months**

-The average LTV of the **20% of those who unsubscribed is $4811 and their tenure is near 50 months**

## 2.What kind of services they subscribed when they were still a customer

In [185]:
# under 80th percentile
total_charges_under80.PhoneService.value_counts()

Yes    1334
No      161
Name: PhoneService, dtype: int64

In [186]:
# Shape(rows,columns) of total_charges_under80
total_charges_under80.shape

(1495, 25)

In [187]:
# Under 80 (percentage)
total_charges_under80.PhoneService.value_counts()  / total_charges_under80.shape[0]

Yes    0.892308
No     0.107692
Name: PhoneService, dtype: float64

In [188]:
#Above 80 (percentage)
total_charges_above80.PhoneService.value_counts()  / total_charges_above80.shape[0]

Yes    0.975936
No     0.024064
Name: PhoneService, dtype: float64

**For the 20% (high-LTV) customers, only 2% did not use the phone service. 
On the other hand, for the customers in the 80% (low-LTV) group, 11% did not use the phone service. That is around 5 times more as the top 20%**

### Multiple lines

In [189]:
# For Total Charges under 80 with multiple lines

total_charges_under80_use_phone = total_charges_under80.query("PhoneService=='Yes'")

total_charges_under80_use_phone.MultipleLines.value_counts() / total_charges_under80_use_phone.shape[0]

No     0.590705
Yes    0.409295
Name: MultipleLines, dtype: float64

In [190]:
# For Total Charges above 80 with multiple lines

total_charges_above80_use_phone = total_charges_above80.query("PhoneService=='Yes'")

total_charges_above80_use_phone.MultipleLines.value_counts() / total_charges_above80_use_phone.shape[0]

Yes    0.832877
No     0.167123
Name: MultipleLines, dtype: float64

**The top 20% of LTV who used the phone service, 84% used multiple lines, which is 2 times as much proportion of people in the low 80% LTV that is around 40%**

**Having multiple lines can bring you closer to be in top 20% group**

### Internet service

In [191]:
# Under 80 (percentage)
total_charges_under80.InternetService.value_counts()  / total_charges_under80.shape[0]

Fiber optic    0.640803
DSL            0.283612
No             0.075585
Name: InternetService, dtype: float64

In [192]:
# After 80 (percentage)
total_charges_above80.InternetService.value_counts()  / total_charges_above80.shape[0]

Fiber optic    0.906417
DSL            0.093583
Name: InternetService, dtype: float64

- All the people in the high 20% LTV used internet service. 
- On the other hand 8% of the people in the low 80th percentile of the LTV did not have internet connection. 
- Also, in the top 20% (high LTV) 90% of them had fiber optic, in comparison to only 64% in the low LTV. 
- So as a company if we want to increase the LTV of customers, we should recommend using internet and Fiber Optic connection

## Other services

In [193]:
import numpy as np

In [194]:
total_charges_under80_use_internet = total_charges_under80.query('InternetService!="No"')
proportion_internet_sub_service_under80 = np.array([total_charges_under80_use_internet.query('OnlineSecurity=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('TechSupport=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('OnlineBackup=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('DeviceProtection=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('StreamingTV=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0],
total_charges_under80_use_internet.query('StreamingMovies=="Yes"').shape[0]/total_charges_under80_use_internet.shape[0]])

In [195]:
print("Total_charges_under80_use_internet")
print()
print("OnlineSecurity-Yes  ",round(proportion_internet_sub_service_under80[0],2),"%")
print("TechSupport-Yes     ",round(proportion_internet_sub_service_under80[1],2),"%")
print("OnlineBackup-Yes    ",round(proportion_internet_sub_service_under80[2],2),"%")
print("DeviceProtection-Yes",round(proportion_internet_sub_service_under80[3],2),"%")
print("StreamingTV-Yes     ",round(proportion_internet_sub_service_under80[4],2),"%")
print("StreamingMovies-Yes ",round(proportion_internet_sub_service_under80[5],2),"%")

Total_charges_under80_use_internet

OnlineSecurity-Yes   0.13 %
TechSupport-Yes      0.14 %
OnlineBackup-Yes     0.23 %
DeviceProtection-Yes 0.24 %
StreamingTV-Yes      0.38 %
StreamingMovies-Yes  0.38 %


In [196]:
total_charges_above80_use_internet = total_charges_above80.query('InternetService!="No"')
proportion_internet_sub_service_above80 = np.array([total_charges_above80_use_internet.query('OnlineSecurity=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('TechSupport=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('OnlineBackup=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('DeviceProtection=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('StreamingTV=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0],
total_charges_above80_use_internet.query('StreamingMovies=="Yes"').shape[0]/total_charges_above80_use_internet.shape[0]])

In [197]:
print("Total_charges_above80_use_internet")
print()
print("OnlineSecurity-Yes  ",round(proportion_internet_sub_service_above80[0],2),"%")
print("TechSupport-Yes     ",round(proportion_internet_sub_service_above80[1],2),"%")
print("OnlineBackup-Yes    ",round(proportion_internet_sub_service_above80[2],2),"%")
print("DeviceProtection-Yes",round(proportion_internet_sub_service_above80[3],2),"%")
print("StreamingTV-Yes     ",round(proportion_internet_sub_service_above80[4],2),"%")
print("StreamingMovies-Yes ",round(proportion_internet_sub_service_above80[5],2),"%")

Total_charges_above80_use_internet

OnlineSecurity-Yes   0.3 %
TechSupport-Yes      0.33 %
OnlineBackup-Yes     0.56 %
DeviceProtection-Yes 0.59 %
StreamingTV-Yes      0.78 %
StreamingMovies-Yes  0.78 %


- In the top 20% LTV, both streaming movies and streaming tv percentages are around 80%, in comparison in the low 80% LTV they are around 40%.
- Device protection and online backup for the top 20% LTV are around 55%, in comparison to only around 22% in the low LTV. 
- As a company we would like to sell to customers streaming TV and movies and also online backup and device protection if we want to increase LTV.

## 3. For those who churned what is the proportion of each kind of contract?

In [198]:
# Under 80 (percentage)
.
total_charges_under80.Contract.value_counts() / total_charges_under80.shape[0]

Month-to-month    0.959197
One year          0.036789
Two year          0.004013
Name: Contract, dtype: float64

In [199]:
# Above 80 (percentage)
total_charges_above80.Contract.value_counts() / total_charges_above80.shape[0]

Month-to-month    0.590909
One year          0.296791
Two year          0.112299
Name: Contract, dtype: float64

## 4. For those who did not churn what is the proportion of each kind of contract?

In [209]:
import numpy as np

In [ ]:
churn_df= pd.read_sql("select * from churn_all where Churn ='Yes'",conn)

In [210]:
## Extract those who subscribed the services (Churn = No)

paying_df= pd.read_sql("select * from churn_all where Churn = 'No'", conn)

In [215]:
churn_df["TotalCharges"].describe()

count    1869.000000
mean     1531.796094
std      1890.822994
min        18.850000
25%       134.500000
50%       703.550000
75%      2331.300000
max      8684.800000
Name: TotalCharges, dtype: float64

In [216]:
paying_df["TotalCharges"].describe()

count     5174
unique    4966
top           
freq        11
Name: TotalCharges, dtype: object

In [213]:
# Find the 80th percentile of the data in TotalCharges
paying_df.TotalCharges.quantile(0.8)

TypeError: '<' not supported between instances of 'str' and 'float'

In [ ]:
# Divide the data by the 80th percentile of the data from the TotalCharges variable
paying_total_charges_under80 = paying_df.query("TotalCharges <= 4895.8")
paying_total_charges_above80 = paying_df.query("TotalCharges > 4895.8")

In [ ]:
# Under 80 (percentage)
paying_total_charges_under80.Contract.value_counts() / paying_total_charges_under80.shape[0]

In [ ]:
# Above 80 (percentage)
paying_total_charges_above80.Contract.value_counts() / paying_total_charges_above80.shape[0]

- If we want to stop Churning we would like to sell contracts to people for at least one year, and much better for 2 years (Conjecture).
- We have 57% of the people that stayed in the company in the top 20% LTV with 2 year contracts, and only 11% of the people in the top 20% had a 2 year contract and left the company.
- 95% of the people that left the company and are in the low 80% have a month-to-month contract, in contrast to only 50% of the people in the low 80% that stayed in the company. We can say that having a month-to-month contract is one of the reasons that people are leaving the company (conjecture).

In [217]:
conn.close()

The approach that we followed for the analysis is described here:

- We began by creating the churn database using all the csv files that were available to us.

- Next, I created an ERD using dbdiagram.io and then checked whether the database satisfied the properties of an RDBMS by validating the normalization rules.

- Then we opened the SQL browser and performed some **data exploration and analysis on the churn data**.

- Further, we moved to Python, **where we removed missing values and joined all the tables to create a single table with all the data that we have with us**.

- After generating the combined dataframe, we found that there was a significant difference in the LTVs of the top 20% and bottom 80% customers.

- Finally, we **performed a complex analysis to check which factors influence the LTVs of the customers and what are the conjectures that we can draw for pulling people into the high-LTV group**.